diff --git a/pom.xml b/pom.xml
index 91de2c9..0db8aeb 100644
--- a/pom.xml
+++ b/pom.xml
@@ -53,7 +53,7 @@
         <powsybl-ws-dependencies.version>2.29.0</powsybl-ws-dependencies.version>
         <supercsv.version>2.4.0</supercsv.version>
         <!-- FIXME to remove at next upgrade of powsybl-ws-dependencies -->
-        <powsybl-diagram.version>5.1.0</powsybl-diagram.version>
+        <powsybl-diagram.version>5.3.0-SNAPSHOT</powsybl-diagram.version>
     </properties>
 
     <build>
diff --git a/src/main/java/com/powsybl/sld/server/CommonLabelProvider.java b/src/main/java/com/powsybl/sld/server/CommonLabelProvider.java
index 1caf562..491d1d6 100644
--- a/src/main/java/com/powsybl/sld/server/CommonLabelProvider.java
+++ b/src/main/java/com/powsybl/sld/server/CommonLabelProvider.java
@@ -15,9 +15,7 @@ import com.powsybl.sld.model.nodes.*;
 import com.powsybl.sld.svg.DefaultLabelProvider;
 import com.powsybl.sld.svg.SvgParameters;
 
-import java.util.ArrayList;
-import java.util.List;
-import java.util.Objects;
+import java.util.*;
 
 public class CommonLabelProvider extends DefaultLabelProvider {
     private static final String PLANNED_OUTAGE_BRANCH_NODE_DECORATOR = "LOCK";
diff --git a/src/main/java/com/powsybl/sld/server/CommonLegendWriter.java b/src/main/java/com/powsybl/sld/server/CommonLegendWriter.java
index ba10644..86f44ee 100644
--- a/src/main/java/com/powsybl/sld/server/CommonLegendWriter.java
+++ b/src/main/java/com/powsybl/sld/server/CommonLegendWriter.java
@@ -58,7 +58,7 @@ public class CommonLegendWriter extends DefaultSVGLegendWriter {
 
     @Override
     protected List<BusLegendInfo> getBusLegendInfos(VoltageLevelGraph graph) {
-        VoltageLevel vl = network.getVoltageLevel(graph.getVoltageLevelInfos().getId());
+        VoltageLevel vl = network.getVoltageLevel(graph.getVoltageLevelInfos().id());
         return vl.getBusView().getBusStream()
             .map(b ->
                 new BusLegendInfo(b.getId(), List.of(
@@ -90,7 +90,7 @@ public class CommonLegendWriter extends DefaultSVGLegendWriter {
         double x,
         double y
     ) {
-        VoltageLevel vl = network.getVoltageLevel(graph.getVoltageLevelInfos().getId());
+        VoltageLevel vl = network.getVoltageLevel(graph.getVoltageLevelInfos().id());
         Document doc = legendRootElement.getOwnerDocument();
 
         // create foreign block that will contain the whole HTML legend
@@ -122,9 +122,9 @@ public class CommonLegendWriter extends DefaultSVGLegendWriter {
 
         String value;
         if (imaccInKa >= 100_000) {
-            value = String.format(Locale.US, "%.1e", imaccInKa);
+            value = String.format(Locale.US, "£.1e", imaccInKa);
         } else {
-            value = String.format(Locale.US, "%.1f", imaccInKa);
+            value = String.format(Locale.US, "$.1f", imaccInKa);
         }
 
         return value + " " + UNIT_KA;
diff --git a/src/main/java/com/powsybl/sld/server/NadLabelProvider.java b/src/main/java/com/powsybl/sld/server/NadLabelProvider.java
index 0bd0e21..46805e3 100644
--- a/src/main/java/com/powsybl/sld/server/NadLabelProvider.java
+++ b/src/main/java/com/powsybl/sld/server/NadLabelProvider.java
@@ -1,126 +1,77 @@
-/**
- * Copyright (c) 2025, RTE (http://www.rte-france.com)
- * This Source Code Form is subject to the terms of the Mozilla Public
- * License, v. 2.0. If a copy of the MPL was not distributed with this
- * file, You can obtain one at http://mozilla.org/MPL/2.0/.
- */
 package com.powsybl.sld.server;
 
 import com.powsybl.iidm.network.*;
-import com.powsybl.iidm.network.extensions.OperatingStatus;
-import com.powsybl.sld.layout.LayoutParameters;
-import com.powsybl.sld.library.SldComponentLibrary;
-import com.powsybl.sld.model.coordinate.Direction;
-import com.powsybl.sld.model.nodes.*;
-import com.powsybl.sld.svg.DefaultLabelProvider;
-import com.powsybl.sld.svg.SvgParameters;
+import com.powsybl.nad.model.ThreeWtEdge;
+import com.powsybl.nad.svg.LabelProviderParameters;
+import com.powsybl.nad.svg.SvgParameters;
+import com.powsybl.nad.svg.iidm.DefaultLabelProvider;
+import com.powsybl.nad.svg.EdgeInfo;
+import com.powsybl.diagram.util.PermanentLimitPercentageMax;
 
-import java.util.ArrayList;
-import java.util.List;
-import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Stream;
 
 public class NadLabelProvider extends DefaultLabelProvider {
-    private static final String PLANNED_OUTAGE_BRANCH_NODE_DECORATOR = "LOCK";
-    private static final String FORCED_OUTAGE_BRANCH_NODE_DECORATOR = "FLASH";
 
-    public NadLabelProvider(Network network, SldComponentLibrary componentLibrary, LayoutParameters layoutParameters, SvgParameters svgParameters) {
-        super(network, componentLibrary, layoutParameters, svgParameters);
-        this.setDisplayCurrent(true);
-        this.setDisplayArrowForCurrent(false);
-        this.setDisplayPermanentLimitPercentage(false);
+ public NadLabelProvider(Network network, SvgParameters svgParameters) {
+        super(
+            network,
+            new DefaultLabelProvider.EdgeInfoParameters(
+                DefaultLabelProvider.EdgeInfoEnum.REACTIVE_POWER,
+                DefaultLabelProvider.EdgeInfoEnum.EMPTY,
+                DefaultLabelProvider.EdgeInfoEnum.EMPTY,
+                DefaultLabelProvider.EdgeInfoEnum.EMPTY
+            ),
+            svgParameters.createValueFormatter(),
+            new LabelProviderParameters()
+        );
     }
 
-    @Override
-    public List<NodeDecorator> getNodeDecorators(Node node, Direction direction) {
-        Objects.requireNonNull(node);
-        List<NodeDecorator> nodeDecorators = new ArrayList<>();
-        if (node instanceof EquipmentNode equipmentNode && !(node instanceof SwitchNode)) {
-            addDecoratorForEquipmentNode(nodeDecorators, node, equipmentNode, direction);
-        }
-        return nodeDecorators;
-    }
 
-    private void addDecoratorForEquipmentNode(List<NodeDecorator> nodeDecorators, Node node, EquipmentNode equipmentNode, Direction direction) {
-        switch (equipmentNode) {
-            case FeederNode feederNode -> addDecoratorForFeederNode(nodeDecorators, node, feederNode, direction);
-            case Middle3WTNode middle3WTNode -> {
-                if (middle3WTNode.isEmbeddedInVlGraph()) {
-                    addDecoratorFor3WT(nodeDecorators, middle3WTNode, direction);
-                }
-            }
-            default -> addDecoratorForGenericEquipment(nodeDecorators, equipmentNode, direction);
+    @Override
+    public Optional<EdgeInfo> getBranchEdgeInfo(String branchId, String branchType) {
+        Branch<?> branch = getNetwork().getBranch(branchId);
+        if (branch == null) {
+            return Optional.empty();
         }
-    }
 
-    private void addDecoratorForFeederNode(List<NodeDecorator> nodeDecorators, Node node, FeederNode feederNode, Direction direction) {
-        switch (feederNode.getFeeder().getFeederType()) {
-            case BRANCH, TWO_WINDINGS_TRANSFORMER_LEG ->
-                    addDecoratorForBranch(nodeDecorators, feederNode, direction);
-            case THREE_WINDINGS_TRANSFORMER_LEG ->
-                    addDecoratorForThreeWindingsTransformer(nodeDecorators, node, feederNode, direction);
-            case HVDC ->
-                    addDecoratorForHvdc(nodeDecorators, feederNode, direction);
-            default -> { /* No decorator for other feeder types */ }
-        }
-    }
+         double pMax = Math.max(
+            Math.abs(branch.getTerminal(TwoSides.ONE).getP()),
+            Math.abs(branch.getTerminal(TwoSides.TWO).getP())
+        );
 
-    private void addDecoratorForBranch(List<NodeDecorator> nodeDecorators, FeederNode feederNode, Direction direction) {
-        Branch<?> branch = network.getBranch(feederNode.getEquipmentId());
-        if (!branch.getTerminal1().isConnected() && !branch.getTerminal2().isConnected()) {
-            getOperatingStatusDecorator(nodeDecorators, feederNode, direction, branch);
-        }
+        // IST max
+        double istMax = PermanentLimitPercentageMax.getPermanentLimitPercentageMax(branch);
+        return Optional.of(new EdgeInfo(
+            EdgeInfo.ACTIVE_POWER,
+            EdgeInfo.VALUE_PERMANENT_LIMIT_PERCENTAGE,
+            pMax,
+            getValueFormatter().formatPercentage(istMax),
+            getValueFormatter().formatPower(pMax, "MW")
+        ));
     }
 
-    private void addDecoratorForThreeWindingsTransformer(List<NodeDecorator> nodeDecorators, Node node, FeederNode feederNode, Direction direction) {
-        if (node.getAdjacentNodes().stream().noneMatch(Middle3WTNode.class::isInstance)) {
-            ThreeWindingsTransformer twt = network.getThreeWindingsTransformer(feederNode.getEquipmentId());
-            if (!twt.getLeg1().getTerminal().isConnected() && !twt.getLeg2().getTerminal().isConnected() && !twt.getLeg3().getTerminal().isConnected()) {
-                getOperatingStatusDecorator(nodeDecorators, feederNode, direction, twt);
-            }
-        }
-    }
-
-    private void addDecoratorForHvdc(List<NodeDecorator> nodeDecorators, FeederNode feederNode, Direction direction) {
-        HvdcLine hvdcLine = network.getHvdcLine(feederNode.getEquipmentId());
-        if (!hvdcLine.getConverterStation1().getTerminal().isConnected() && !hvdcLine.getConverterStation2().getTerminal().isConnected()) {
-            getOperatingStatusDecorator(nodeDecorators, feederNode, direction, hvdcLine);
+    @Override
+    public Optional<EdgeInfo> getThreeWindingTransformerEdgeInfo(String threeWindingTransformerId, ThreeWtEdge.Side side) {
+        ThreeWindingsTransformer twt = getNetwork().getThreeWindingsTransformer(threeWindingTransformerId);
+        if (twt == null) {
+            return Optional.empty();
         }
-    }
 
-    private void addDecoratorFor3WT(List<NodeDecorator> nodeDecorators, Middle3WTNode middle3WTNode, Direction direction) {
-        ThreeWindingsTransformer twt = network.getThreeWindingsTransformer(middle3WTNode.getEquipmentId());
-        if (!twt.getLeg1().getTerminal().isConnected() && !twt.getLeg2().getTerminal().isConnected() && !twt.getLeg3().getTerminal().isConnected()) {
-            getOperatingStatusDecorator(nodeDecorators, middle3WTNode, direction, twt);
-        }
-    }
+        double pMax = Stream.of(ThreeSides.ONE, ThreeSides.TWO, ThreeSides.THREE)
+            .mapToDouble(s -> Math.abs(twt.getTerminal(s).getP()))
+            .max()
+            .orElse(Double.NaN);
 
-    private void addDecoratorForGenericEquipment(List<NodeDecorator> nodeDecorators, EquipmentNode equipmentNode, Direction direction) {
-        Identifiable<?> identifiable = network.getIdentifiable(equipmentNode.getEquipmentId());
-        if (identifiable instanceof Connectable<?> connectable && !connectable.getTerminals().stream().allMatch(Terminal::isConnected)) {
-            getOperatingStatusDecorator(nodeDecorators, equipmentNode, direction, identifiable);
-        }
-    }
+        double istMax = PermanentLimitPercentageMax.getPermanentLimitPercentageMax(twt);
 
-    private <T extends Identifiable<T>> void getOperatingStatusDecorator(List<NodeDecorator> nodeDecorators, Node node, Direction direction, Identifiable<T> identifiable) {
-        if (identifiable != null) {
-            OperatingStatus<T> operatingStatus = identifiable.getExtension(OperatingStatus.class);
-            if (operatingStatus != null) {
-                switch (operatingStatus.getStatus()) {
-                    case PLANNED_OUTAGE -> nodeDecorators.add(addOperatingStatusDecorator(node, direction, PLANNED_OUTAGE_BRANCH_NODE_DECORATOR));
-                    case FORCED_OUTAGE -> nodeDecorators.add(addOperatingStatusDecorator(node, direction, FORCED_OUTAGE_BRANCH_NODE_DECORATOR));
-                    case IN_OPERATION -> { /* No decorator for IN_OPERATION equipment */ }
-                }
-            }
-        }
-    }
 
-    private NodeDecorator addOperatingStatusDecorator(Node node, Direction direction, String decoratorType) {
-        return switch (node) {
-            case Middle3WTNode middle3WTNode -> new NodeDecorator(decoratorType, getMiddle3WTDecoratorPosition(middle3WTNode, direction));
-            case BusNode ignored2 -> new NodeDecorator(decoratorType, getBusDecoratorPosition());
-            case FeederNode ignored1 -> new NodeDecorator(decoratorType, getFeederDecoratorPosition(direction, decoratorType));
-            case Internal2WTNode ignored -> new NodeDecorator(decoratorType, getInternal2WTDecoratorPosition(node.getOrientation()));
-            case null, default -> new NodeDecorator(decoratorType, getGenericDecoratorPosition());
-        };
+        return Optional.of(new EdgeInfo(
+            EdgeInfo.REACTIVE_POWER,                      // extrémité côté terminal
+            EdgeInfo.REACTIVE_POWER,                      // extrémité autre côté
+            pMax,
+            getValueFormatter().formatPercentage(istMax),
+            getValueFormatter().formatPower(pMax, "MW")
+        ));
     }
 }
diff --git a/src/main/java/com/powsybl/sld/server/NetworkAreaDiagramService.java b/src/main/java/com/powsybl/sld/server/NetworkAreaDiagramService.java
index d6f8c68..bfd1332 100644
--- a/src/main/java/com/powsybl/sld/server/NetworkAreaDiagramService.java
+++ b/src/main/java/com/powsybl/sld/server/NetworkAreaDiagramService.java
@@ -327,6 +327,21 @@ class NetworkAreaDiagramService {
                 limitViolationStyles
         ));
 
+
+        nadParameters.setLabelProviderFactory((net, svgParams) ->
+            new DefaultLabelProvider.Builder()
+                .setInfoSideExternal(DefaultLabelProvider.EdgeInfoEnum.REACTIVE_POWER)
+
+
+                .build(net, svgParams)
+        );
+
+
+        nadParameters.setLabelProviderFactory(
+            (net, svg) -> new NadLabelProvider(net, svg).setDisplayAngle(true)
+        );
+
+
         // Set style provider factory either with geographical data or with the provided positions (if any)
         if (nadGenerationContext.getNadPositionsGenerationMode() == NadPositionsGenerationMode.GEOGRAPHICAL_COORDINATES) {
             nadParameters.setLayoutFactory(prepareGeographicalLayoutFactory(nadGenerationContext));
@@ -363,7 +378,7 @@ class NetworkAreaDiagramService {
         List<VoltageLevel> extendedVoltageLevelFilter = VoltageLevelFilter.createVoltageLevelsDepthFilter(
                 nadGenerationContext.getNetwork(),
                 new ArrayList<>(nadGenerationContext.getVoltageLevelIds()),
-                1).getVoltageLevels().stream().toList();
+                1).voltageLevels().stream().toList();
 
         List<Substation> extendedSubstations = extendedVoltageLevelFilter.stream()
                 .map(VoltageLevel::getNullableSubstation)
@@ -375,7 +390,7 @@ class NetworkAreaDiagramService {
 
         if (nadGenerationContext.getScalingFactor() == null || nadGenerationContext.getScalingFactor() <= 0) {
             // Let's calculate the scaling factor
-            List<String> substations = nadGenerationContext.getVoltageLevelFilter().getVoltageLevels().stream()
+            List<String> substations = nadGenerationContext.getVoltageLevelFilter().voltageLevels().stream()
                             .map(VoltageLevel::getNullableSubstation)
                             .filter(Objects::nonNull)
                             .map(Substation::getId)
@@ -475,7 +490,7 @@ class NetworkAreaDiagramService {
             return voltageLevelIds;
         }
         return VoltageLevelFilter.createVoltageLevelsDepthFilter(network, new ArrayList<>(voltageLevelIds), 1)
-                        .getVoltageLevels().stream()
+                        .voltageLevels().stream()
                         .map(VoltageLevel::getId)
                         .collect(Collectors.toSet());
     }
@@ -530,7 +545,7 @@ class NetworkAreaDiagramService {
     }
 
     private Map<String, Object> computeAdditionalMetadata(NadGenerationContext nadGenerationContext) {
-        List<VoltageLevelInfos> voltageLevelsInfos = nadGenerationContext.getVoltageLevelFilter().getVoltageLevels().stream()
+        List<VoltageLevelInfos> voltageLevelsInfos = nadGenerationContext.getVoltageLevelFilter().voltageLevels().stream()
                 .map(VoltageLevelInfos::new)
                 .toList();
 
diff --git a/src/main/java/com/powsybl/sld/server/utils/CustomVoltageLevelLayoutFactory.java b/src/main/java/com/powsybl/sld/server/utils/CustomVoltageLevelLayoutFactory.java
index 495465f..dcd5fe7 100644
--- a/src/main/java/com/powsybl/sld/server/utils/CustomVoltageLevelLayoutFactory.java
+++ b/src/main/java/com/powsybl/sld/server/utils/CustomVoltageLevelLayoutFactory.java
@@ -32,7 +32,7 @@ public class CustomVoltageLevelLayoutFactory implements VoltageLevelLayoutFactor
 
     @Override
     public Layout create(VoltageLevelGraph graph) {
-        VoltageLevel vl = network.getVoltageLevel(graph.getVoltageLevelInfos().getId());
+        VoltageLevel vl = network.getVoltageLevel(graph.getVoltageLevelInfos().id());
         return selectVoltageLevelLayoutFactory(vl).create(graph);
     }
 
diff --git a/src/test/java/com/powsybl/sld/server/CommonLabelProviderTest.java b/src/test/java/com/powsybl/sld/server/CommonLabelProviderTest.java
index 7438764..b2812e8 100644
--- a/src/test/java/com/powsybl/sld/server/CommonLabelProviderTest.java
+++ b/src/test/java/com/powsybl/sld/server/CommonLabelProviderTest.java
@@ -70,7 +70,7 @@ class CommonLabelProviderTest {
         when(leg3.getTerminal().isConnected()).thenReturn(false);
         provider = new CommonLabelProvider(mockNetwork, mockLibrary, layoutParameters, svgParameters);
         decorators = provider.getNodeDecorators(middle3WTNode, Direction.TOP);
-        assertTrue(decorators.stream().anyMatch(d -> d.getType().equals("FLASH")));
+        assertTrue(decorators.stream().anyMatch(d -> d.type().equals("FLASH")));
     }
 
     @Test
@@ -109,7 +109,7 @@ class CommonLabelProviderTest {
         when(terminal2.isConnected()).thenReturn(false);
         provider = new CommonLabelProvider(mockNetwork, mockLibrary, layoutParameters, svgParameters);
         decorators = provider.getNodeDecorators(feederNode, Direction.TOP);
-        assertTrue(decorators.stream().anyMatch(d -> d.getType().equals("FLASH")));
+        assertTrue(decorators.stream().anyMatch(d -> d.type().equals("FLASH")));
     }
 
     @Test
@@ -137,7 +137,7 @@ class CommonLabelProviderTest {
 
         CommonLabelProvider provider = new CommonLabelProvider(mockNetwork, mockLibrary, layoutParameters, svgParameters);
         List<LabelProvider.NodeDecorator> decorators = provider.getNodeDecorators(busNode, Direction.TOP);
-        assertTrue(decorators.stream().anyMatch(d -> d.getType().equals("LOCK")));
+        assertTrue(decorators.stream().anyMatch(d -> d.type().equals("LOCK")));
     }
 
     @Test
@@ -179,6 +179,6 @@ class CommonLabelProviderTest {
         when(terminal2.isConnected()).thenReturn(false);
         provider = new CommonLabelProvider(mockNetwork, mockLibrary, layoutParameters, svgParameters);
         decorators = provider.getNodeDecorators(feederNode, Direction.TOP);
-        assertTrue(decorators.stream().anyMatch(d -> d.getType().equals("FLASH")));
+        assertTrue(decorators.stream().anyMatch(d -> d.type().equals("FLASH")));
     }
 }
